text/0000-otlp-proto-version-identifier.md

```diff
@@ -0,0 +1,31 @@
+# Include OTLP proto version identifier in requests
+
+Adds OTLP request identifer to descrive the OTLP version used to create the binary or JSON payload.
+
+## Motivation
+
+It is useful to know the version of the OTLP protobuf definition used to create the binary or JSON.
+
+This would enable routing to different services by inspecting the header value. This is especially useful for organisations that wish to support multiple versions where the OTLP proto definitions have changed. A currently used alternative is to append the version to the path, but this is clumsy and requires additional configuration when setting up the exporter.
+
+## Explanation
+
+Currently it is not possible to know the version of the OTLP proto definition before receving it. This is challening because there have been changes to the OTLP proto specification that are breaking.
+
+This also aids organisations that receive multiple streams of OTLP data and be able to monitor and encourage users to upgrade to newer versions as appropriate.
+
+## Internal details
+
+* Add gRPC metadta entry or HTTP request header entry with the OTLP version used to create the binary or JSON payload
+
+For example, we could use the key `x-otlp-version` and the string value of the OTLP proto version in the form "0.17.0".
+
+## Trade-offs and mitigations
+
+Currently releaed SDKs and collector versions will not include these identifiers.
+
+## Prior art and alternatives
```

Comment:
Previously people asked for the version number in the payload a few times. Can you consider this alternate and why in the header and not in the payload (or vice versa)?

Reply:
Add considered request body resource attribute in alternatives. Headers offer more flexibility in routing the request as they body does not need to be decoded to get the value.

Reply:
Hey @tigrannajaryan - sorry for delay in replying.

Encoding the OTLP version within the payload doesn't feel like a good solution as we'd be forcing consumers who do care about the encoding to decode twice, first to get the OTLP version and again to decode using the appropriate OTLP handler. OTLP request object can be large and it's very inefficient the whole rquest to potentially throw away. Not only that, many gRPC handler frameworks don't even offer you the raw binary payload and handle the decoding for you.

For the use case of OTLP requests that have been serialised to disk (eg binary / JSON), do we have some way to describe the metadata associated to the request payload when it was received?

For example the structure could be something like this in JSON:

```json
{
  "metadata": {
    "otlp-version": "0.17.0",
  },
  "request": {
    ... JSON for request
  }
```

Is there a stable format request data to use when serialising to disk?

Reply:
You're right - the original purpose of the OTLP version being available was only to be used for informational reasons (eg identify users who send which versions and promote them to upgrade as appropriate).

However, we are in a situation where there have seen some breaking changes in the 0.x versions of the proto definitions, particularly around the fairly aggressive removing of deprecated types / fields. A good example of this is the deprecated Int/Float metrics that were removed in proto version 0.12 and new fields has been added to ScopeXXX in 0.17. If deprecated fields had been left in the proto files until the 1.0 release, they likely would not have caused this problem.

I also believe the use-case originally shared where some receivers wish to route requests based on the OTLP version is still valid and would be best served via a gRPC metadata entry or HTTP header. This would be possible if the OTLP version was part of the request payload, but would require the routing process to decode the whole body to get the version and potentially route to handler which would then decode again.

Why do you feel having the version both in metadata/header and request body field be an issue?

Reply:
To follow-up on this;

I think the OTLP version should be treated as metadata for the given request and should be stored separately from the payload data. For example, if you're storing requests for playback at a later time, you'd likely want to record other useful information such as the sender (eg user agent header) and the time it was received. I think the OTLP version also falls into the category of useful info, that helps you understand and process the request payload.

Reply:
@tigrannajaryan @jmacd @bogdandrutu Any further feedback on this?